woocommerce/wc-admin
# Adding Slotfill extension components for remote payments

Fixes #6853 

This is a starting point for a potential Slotfill solution for remote payments. It's one potential approach, and there is freedom here to be fairly strategic with how we place/configure Slotfills. I've tried to do so in a way that maximized capability while keeping it fairly simple to use. 

I've also created a `<SettingsForm />` component that accepts a variety of props including `fields`. The fields should match the format as those stored in the settings API (or an array of the values in that object). Right now I'm just filtering these to only those that "appear" in the remote config, but this will be altered to tap into @joshuatf 's revisions to the API that will determine essential fields. Right now it only supports text, password, checkbox and select type fields. It's fairly immature so far, but works well for this use case and is ripe for further extension. Open for any feedback on what we want to include/change on this in the context of this PR. 
```js
<SettingsForm
	fields={ fields }
	isBusy={ isOptionsRequesting }
	onSubmit={ updateSettings }
	onButtonClick={ () => recordConnectStartEvent( key ) }
	buttonLabel={ __( 'Proceed', 'woocommerce-admin' ) }
	validate={ validate }
	{ ...props }
/>
```

Details on the Slotfills:
#### `<WooRemotePayment />`
This is a Slot over the entire `<Stepper />` that can ultimately allow for customization of the entire setup/configuration process, including adding steps, etc. 
```js
<WooRemotePayment.Slot
	defaultStepper={ DefaultStepper }
	defaultInstallStep={ installStep }
	defaultConnectStep={ connectStep }
	id={ key }
/>
```

It also passes some helpful props:
**defaultStepper**: The default, pre-configured stepper component that can be customized by overriding specific props.
**defaultInstallStep**: The default installation step, in case they just want to add steps, and plug in the defaults.
**defaultConnectStep**: Same as above, for the connect step.

#### `<WooRemotePaymentSettings />`
We're using the new `<Settings />` component with `<PaymentConnect />`, and this slot will be used as an alternative. 
```js
<WooRemotePaymentSettings.Slot
	defaultSettings={ DefaultSettings }
	defaultUpdate={ updateSettings }
	markConfigured={ () => markConfigured( key ) }
	id={ key }
/>
```
Helpful props:
**defaultSettings**: The default settings component, with can have individual props overridden.
**defaultUpdate**: The default update function passed to `<Settings />`.
**markConfigured**: The callback to mark the method as configured, but no need to pass the method key.


### Screenshots

![image](https://user-images.githubusercontent.com/444632/117180663-441fed80-ad89-11eb-9592-5e2ad4d0c8da.png)

### Detailed test instructions:

- Checkout branch
- Enable remote payments feature flag
- You can checkout [my fork of the remote payment tester](https://github.com/joelclimbsthings/remote-payments-tester) with added front-end tooling (I can merge this into the original if we want).
- Activate the remote payment tester, and run `npm install` and `npm start`
- Navigate to the payment task and click on one of the sample configs (payfast, stripe). 
- You should see the fields as pictured in the screenshot. These are built from the settings API, but limited by the sample configuration provided by the tester plugin (again, this will change a bit after enhancements to the API).
- Notice the custom headings. These are inserted via Slotfill, and the `update` function is overridden as well to just do a `console.info` from the tester plugin.
- Poke around with the demo Slotfills in the `src/index.js` of the tester plugin, and temporarily remove them to ensure that the default functionality steps in as it should.
